cpp/arcticdb/version/symbol_list.cpp

```diff
@@ -5,83 +5,104 @@
-                    write_symbols(store, symbols, compaction_id, creation_ts).get();
-                    delete_keys(store, all_keys);
-                } else {
-                    SYMBOL_LIST_RUNTIME_LOG("Not writing symbols as another write is in progress");
-                }
-            } catch (const storage::PermissionException& ex) {
-                SYMBOL_LIST_RUNTIME_LOG("Can't compact symbol list in read-only mode");
-            } catch (const std::system_error& ex) {
-                SYMBOL_LIST_RUNTIME_LOG("Caught error in symbol list write: {}", ex.what());
-            }
-            // Acceptable to swallow the exceptions above and return symbols because it is loaded from the source
-            // (version keys) without involving the (inaccessible) symbol list cache:
-            return symbols;
-        }
-        else {
-            if(all_keys.size() > max_delta_ && !no_compaction) {
-                SYMBOL_LIST_RUNTIME_LOG("Symbol chain too long, doing compaction");
-                return compact(store, all_keys);
-            }
-            else {
-                auto compaction_and_deltas = KeyVector{all_keys.begin() + maybe_last_compaction.value_or(0), all_keys.end()};
-                return load_from_storage(store, compaction_and_deltas);
-            }
-        }
+struct SymbolList::LoadResult {
+    mutable KeyVector symbol_list_keys;
+    /** The last CompactionId key in symbol_list_keys, if any. */
+    std::optional<KeyVectorItr> maybe_last_compaction;
+    mutable CollectionType symbols;
+    arcticdb::timestamp timestamp;
+
+    KeyVector&& detach_symbol_list_keys() const { return std::move(symbol_list_keys); }
+    CollectionType&& detach_symbols() const { return std::move(symbols); }
+};
+
+SymbolList::LoadResult SymbolList::attempt_load(const std::shared_ptr<Store>& store) {
+    SYMBOL_LIST_RUNTIME_LOG("Symbol list load attempt");
+    LoadResult load_result;
+    // See RFC 006 for the explanation of each step below.
+    // Step 0
+    load_result.symbol_list_keys = get_all_symbol_list_keys(store);
+    // Step 1
+    load_result.maybe_last_compaction = last_compaction(load_result.symbol_list_keys);
+
+    if (load_result.maybe_last_compaction) {
+        load_result.timestamp = load_result.symbol_list_keys.rbegin()->creation_ts(); // Per step 5
+        load_result.symbols = load_from_symbol_list_keys(store,
+                {*load_result.maybe_last_compaction, load_result.symbol_list_keys.cend()});
+    } else {
+        load_result.timestamp = store->current_timestamp();
+        load_result.symbols = load_from_version_keys(store);
     }
+    return load_result;
+}
 
-    SymbolList::CollectionType SymbolList::compact(std::shared_ptr<Store> store, const std::vector<AtomKey>& symbol_keys) {
+SymbolList::CollectionType SymbolList::load(const std::shared_ptr<Store>& store, bool no_compaction) {
+    // Steps 0-1
+    const LoadResult load_result = ExponentialBackoff<std::runtime_error>(100, 2000)
```

Comment:
Do we actually want to catch runtime_error? What error conditions is it legitimate to ignore?

Reply:
There are two classes of issues to consider:

* Some `Storage` don't throw the correct exception type in all cases
* We will need to check if symbol list code itself need this

Given this implementation has been validated internally, I am more inclined to release this first and then improve.

Reply:
I tried to fix the storage when we did this change internally, but it involves many layers of storage abstraction, so requires significant effort to check and test all the downstream code.

This symbol list change has been verified internally and is known to improve things, so I want it decoupled from any potential change that can break things. I have logged #447.